cmd/snap-update-ns/change.go

```diff
@@ -57,28 +58,180 @@ func (c Change) String() string {
 // changePerform is Change.Perform that can be mocked for testing.
 var changePerform func(*Change) ([]*Change, error)
 
-// changePerformImpl is the real implementation of Change.Perform
-func changePerformImpl(c *Change) ([]*Change, error) {
-	if c.Action == Mount {
-		const (
-			mode = 0755
-			uid  = 0
-			gid  = 0
-		)
-		// Create target mount directory if needed.
-		if err := ensureMountPoint(c.Entry.Dir, mode, uid, gid); err != nil {
-			return nil, err
+func (c *Change) createInodes(path, kind string, pokeHoles bool) ([]*Change, error) {
+	var err error
+	var changes []*Change
+
+	// In case we need to create something, some constants.
+	const (
+		mode = 0755
+		uid  = 0
+		gid  = 0
+	)
+
+	// If the element doesn't exist we can attempt to create it.  We will
+	// create the parent directory and then the final element relative to it.
+	// The traversed space may be writable so we just try to create things
+	// first.
+
+	// TODO: refactor this, if possible, with inspection and pre-emptive
+	// creation after the current release ships. This should be possible but
+	// will affect tests heavily (churn, not safe before release).
+retry:
+	switch kind {
+	case "":
+		err = secureMkdirAll(path, mode, uid, gid)
+	case "file":
+		err = secureMkfileAll(path, mode, uid, gid)
+	case "symlink":
+		err = secureMkdirAll(path, mode, uid, gid)
+		if err == nil {
+			// Normally symlinks would be created later but calling
+			// c.lowLevelPerform now lets us check if the medium is
+			// writable and take advantage of the code below.
+			err = c.lowLevelPerform()
```

Comment:
Why do we do that for symlinks but not for everything else? Still feels like functions are crossing into each other.

Reply:
Because mount and unmount rely on existing inodes while symlinks need a new inode. When we create a new inode we can actually realize this place is read-only. This is why it is here.

Reply:
This got refactored away.